### PR TITLE
Screenschotfolderbutton #1642

### LINKED ISSUE #1642: Add button in options to open screenshots folder
to make them easier to find

### PR DESCRIPTION
**Pull request by Paulpicotin**
closes #1642

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
You can force push over existing PRs. So there is no need to create a new PR to even replace every single commit in the PR. If you can't push normally use force pushing to force overwrite commits in a PR.

File: src/engine/Settings.cs [hunk 1]
```diff
@@ -185,6 +185,8 @@ private Settings()
     /// </summary>
     public SettingValue<string> CustomUsername { get; set; } = new SettingValue<string>(null);
 
+    public SettingValue<bool> OpenScreenShotFolder { get; set; } = new SettingValue<bool>(false);
```
Comment (hhyyrylainen) [inline src/engine/Settings.cs:188]: Why did you add a new setting here? A button just for opening the screenshot folder doesn't require any state?

File: src/engine/Settings.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿using Godot;
+using Godot;
```
Comment (hhyyrylainen) [inline src/engine/Settings.cs:1]: Did you add a BOM here? The files should be without that.

File: src/general/OptionsMenu.cs [hunk 3]
```diff
@@ -394,6 +400,7 @@ public void ApplySettingsToControls(Settings settings)
             settings.CustomUsername :
             Environment.UserName;
         customUsername.Editable = settings.CustomUsernameEnabled;
+        goToScreenshotFolder.Pressed = settings.OpenScreenShotFolder;
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:403]: Shouldn't the button for the screenshot folder opening just be a normal button (that doesn't have pressed down state)?

File: src/general/OptionsMenu.cs [hunk 4]
```diff
@@ -144,6 +144,9 @@ public class OptionsMenu : Control
     [Export]
     public NodePath CustomUsernamePath;
 
+    [Export]
+    public NodePath ScreenshotFolderPath;
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:148]: I think you can just remove this and goToScreenshotFolder field

File: src/general/OptionsMenu.cs [hunk 5]
```diff
@@ -973,4 +980,12 @@ private void OnCustomUsernameTextChanged(string text)
 
         UpdateResetSaveButtonState();
     }
+
+    private void OnOpenScreenshotFolder()
+    {
+        GUICommon.Instance.PlayButtonPressSound();
+
+        OS.ShellOpen(ProjectSettings.GlobalizePath(Constants.SCREENSHOT_FOLDER));
+        UpdateResetSaveButtonState();
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:989]: Isn't it unnecessary to call this as this button doesn't change the options state?

File: src/general/OptionsMenu.tscn [hunk 6]
```diff
@@ -825,6 +826,13 @@ margin_right = 511.0
 margin_bottom = 27.0
 rect_min_size = Vector2( 300, 0 )
 
+[node name="ScreenShot folder" type="Button" parent="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer"]
+margin_top = 349.0
+margin_right = 511.0
+margin_bottom = 379.0
+action_mode = 0
+text = "Screenshot Folder"
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.tscn:834]: I think the button should say "Open Screenshot Folder"

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/general/OptionsMenu.tscn (hunk 6 above)
Comment (Paulpicotin) [reply src/general/OptionsMenu.tscn:834]: ok I changed that

File: src/general/OptionsMenu.cs (hunk 4 above)
Comment (Paulpicotin) [reply src/general/OptionsMenu.cs:148]: I remove all of the field so

File: src/engine/Settings.cs (hunk 2 above)
Comment (Paulpicotin) [reply src/engine/Settings.cs:1]: I don't know about that but when I remove using Godot;  it gave me an error

Comment (hhyyrylainen) [reply src/engine/Settings.cs:1]: BOM is an invisible character that a file can begin with.
https://en.wikipedia.org/wiki/Byte_order_mark

When a diff shows no changed visible characters, and it is the first line, my first guess is that your editor added a BOM, our source files don't use that so you need to edit the file to remove it.

Comment (hhyyrylainen) [reply src/engine/Settings.cs:1]: Opened an issue to make an automatic check for this: https://github.com/Revolutionary-Games/Thrive/issues/1786

File: src/general/OptionsMenu.cs (hunk 5 above)
Comment (hhyyrylainen) [reply src/general/OptionsMenu.cs:982]: I think this is the last problem with this.
I merged master into this branch, that apparently removed the one BOM change? Next time you work on this you should pull first before starting working.

File: src/general/OptionsMenu.tscn [hunk 7]
```diff
@@ -825,6 +825,13 @@ margin_right = 511.0
 margin_bottom = 27.0
 rect_min_size = Vector2( 300, 0 )
 
+[node name="ScreenShot folder" type="Button" parent="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer"]
+margin_top = 349.0
+margin_right = 511.0
+margin_bottom = 379.0
+action_mode = 0
+text = "Open Screenshot Folder"
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.tscn:833]: This needs to be changed to use the new translations system.

File: src/general/OptionsMenu.cs [hunk 8]
```diff
@@ -973,4 +973,12 @@ private void OnCustomUsernameTextChanged(string text)
...
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:982]: And this `UpdateResetSaveButtonState` should still be deleted.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/general/OptionsMenu.tscn (hunk 7 above)
Comment (Paulpicotin) [reply src/general/OptionsMenu.tscn:833]: What am I suppose to do there?

Comment (hhyyrylainen) [reply src/general/OptionsMenu.tscn:833]: https://github.com/Revolutionary-Games/Thrive/blob/master/doc/working_with_translations.md#adding-new-content-into-the-game-with-translation-in-mind

File: src/engine/Settings.cs [hunk 9]
```diff
@@ -214,6 +214,430 @@ private Settings()
...
+
+            // Since the properties we want to copy are SettingValue generics we use the IAssignableSetting
+            // interface and AssignFrom method to convert the property to the correct concrete class.
+            var setting = (IAssignableSetting)property.GetValue(this);
+
+            setting.AssignFrom(property.GetValue(settings));
+        }
+    }
+}
+using Godot;
```
Comment (hhyyrylainen) [inline src/engine/Settings.cs:463]: Looks like a badly gone merge resolve?

File: project.godot [hunk 10]
```diff
@@ -215,7 +215,6 @@ translations=PoolStringArray( "res://locale/en.po", "res://locale/fr.po", "res:/
 [logging]
 
 file_logging/enable_file_logging=true
-file_logging/log_path="user://logs/log.txt"
```
Comment (hhyyrylainen) [inline project.godot:218]: This needs to be reverted.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: project.godot (hunk 10 above)
Comment (Paulpicotin) [reply project.godot:218]: that's just a bad merge cause when i just come back on the code I got this line replace : file_logging/max_log_files=8

**Comment by hhyyrylainen** [comment]
How did you make all of those translations for the button? If they are like google translated or something, they should be rather left blank than guessed.

File: src/general/OptionsMenu.tscn [hunk 11]
```diff
@@ -1187,8 +1196,10 @@ dialog_text = "ERROR_FAILED_TO_SAVE_NEW_SETTINGS"
...
 [connection signal="toggled" from="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer/TutorialsEnabled" to="." method="OnTutorialsEnabledToggled"]
 [connection signal="toggled" from="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer/Cheats" to="." method="OnCheatsToggled"]
+[connection signal="pressed" from="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer/Screeshot Folder" to="." method="OnOpenScreenshotFolder"]
 [connection signal="toggled" from="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer/CustomUsernameEnabled" to="." method="OnCustomUsernameEnabledToggled"]
 [connection signal="text_changed" from="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer/HBoxContainer3/CustomUsername" to="." method="OnCustomUsernameTextChanged"]
+[connection signal="pressed" from="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer/ScreenShot folder" to="." method="OnOpenScreenshotFolder"]
```
Comment (FuJa0815) [inline src/general/OptionsMenu.tscn:1202]: Wrong merge?
The first line has a spelling mistake

File: src/general/OptionsMenu.cs [hunk 12]
```diff
@@ -1034,14 +1034,21 @@ private void BuildInputRebindControls()
...
+    }
+    
     private void OnCustomUsernameEnabledToggled(bool pressed)
     {
         Settings.Instance.CustomUsernameEnabled.Value = pressed;
         customUsername.Editable = pressed;
 
         UpdateResetSaveButtonState();
     }
-
```
Comment (FuJa0815) [inline src/general/OptionsMenu.cs:1044]: Re-add that blank like

File: src/general/OptionsMenu.cs [hunk 13]
```diff
@@ -1034,14 +1034,21 @@ private void BuildInputRebindControls()
         inputGroupList.InitGroupList();
     }
 
+    private void OnOpenScreenshotFolder()
+    {
+        GUICommon.Instance.PlayButtonPressSound();
+
+        OS.ShellOpen(ProjectSettings.GlobalizePath(Constants.OPEN_SCREENSHOT_FOLDER));
+
```
Comment (FuJa0815) [inline src/general/OptionsMenu.cs:1042]: remove that blank line

File: src/engine/Settings.cs [hunk 14]
```diff
@@ -214,6 +214,430 @@ private Settings()
             CustomUsername.Value :
             Environment.UserName;
 
+    public SettingValue<bool> OpenScreenshotFolder {get; set; } = new SettingValue<bool>(false);
```
Comment (FuJa0815) [inline src/engine/Settings.cs:217]:
```suggestion
    public SettingValue<bool> OpenScreenshotFolder { get; set; } = new SettingValue<bool>(false);
```

File: src/engine/Settings.cs [hunk 15]
```diff
@@ -214,6 +214,430 @@ private Settings()
...
+    ///   Show tutorial messages
+    /// </summary>
+    public SettingValue<bool> TutorialsEnabled { get; set; } = new SettingValue<bool>(true);
+
+    /// <summary>
+    ///   When true cheats are enabled
+    /// </summary>
+    public SettingValue<bool> CheatsEnabled { get; set; } = new SettingValue<bool>(false);
+
+    public SettingValue<bool> OpenScreenshotFolder {get; set; } = new SettingValue<bool>(false);
```
Comment (FuJa0815) [inline src/engine/Settings.cs:639]: What happend here?
Looks like you CTRL+C, V'd half of the class. You declare two `Settings` classes in the `Settings.cs`
```suggestion

```

File: src/engine/Settings.cs [hunk 16]
```diff
@@ -563,4 +987,4 @@ private void CopySettings(Settings settings)
             setting.AssignFrom(property.GetValue(settings));
         }
     }
-}
+}
```
Comment (FuJa0815) [inline src/engine/Settings.cs:990]: Add a newline at the end of the file

**Review by FuJa0815 — CHANGES_REQUESTED** [review]

File: src/general/OptionsMenu.cs (hunk 13 above)
Comment (FuJa0815) [reply src/general/OptionsMenu.cs:1042]: the one on line 1042

File: src/engine/Settings.cs (hunk 15 above)
Comment (Paulpicotin) [reply src/engine/Settings.cs:639]: I think that is a bad merge sorry I'm lookin in today sorry for that

File: locale/ko.po [hunk 17]
```diff
@@ -1110,7 +1114,7 @@ msgstr "키 입력을 기본값으로 초기화하시겠습니까?"
 msgid "CLOSE_OPTIONS"
 msgstr "설정을 종료합니까?"
 
-#: ../src/general/OptionsMenu.tscn:1065
+
```
Comment (hhyyrylainen) [inline locale/ko.po:1117]: Unintended deletion?

File: locale/en.po [hunk 18]
```diff
@@ -1095,9 +1095,13 @@ msgstr "Use a custom username"
 msgid "CUSTOM_USERNAME"
 msgstr "Custom Username:"
 
-#: ../src/general/OptionsMenu.tscn:972 ../src/general/PauseMenu.tscn:133
-#: ../src/gui_common/MainMenu.tscn:319 ../src/saving/NewSaveMenu.tscn:109
-#: ../src/saving/SaveManagerGUI.tscn:112
+#: ../src/general/OptionsMenu.tscn:970
+msgid "OPEN_SCREENSHOT_FOLDER"
+msgstr "open screenshot folder"
```
Comment (hhyyrylainen) [inline locale/en.po:1100]: `Open Screenshot Folder`

File: project.godot (hunk 10 above)
Comment (hhyyrylainen) [reply project.godot:249]: this change is still here

File: src/engine/Settings.cs [hunk 19]
```diff
@@ -215,9 +215,9 @@ private Settings()
         CustomUsername.Value != null ?
             CustomUsername.Value :
             Environment.UserName;
+            
 
     public int CloudSimulationWidth => Constants.CLOUD_X_EXTENT / CloudResolution;
-
```
Comment (hhyyrylainen) [inline src/engine/Settings.cs:220]: All changes in this file should be reverted.

File: src/general/OptionsMenu.tscn [hunk 20]
```diff
@@ -193,6 +193,7 @@ custom_constants/margin_left = 15
 custom_constants/margin_bottom = 15
 
 [node name="Graphics" type="Control" parent="CenterContainer/VBoxContainer/PanelContainer/MarginContainer"]
+visible = false
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.tscn:196]: Is this a problematic change? I think it is.
@Kasterisk can probably confirm.

File: src/general/OptionsMenu.tscn [hunk 21]
```diff
@@ -795,7 +796,6 @@ margin_bottom = 32.0
 text = "RESET_INPUTS"
 
 [node name="Misc" type="Control" parent="CenterContainer/VBoxContainer/PanelContainer/MarginContainer"]
-visible = false
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.tscn:798]: Related to the other `visible` change.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: locale/ko.po (hunk 17 above)
Comment (Paulpicotin) [reply locale/ko.po:1117]: yes it's changed

File: project.godot (hunk 10 above)
Comment (Paulpicotin) [reply project.godot:249]: I don't understand ? I am  supposed to replace 
file_logging/log_path="user://logs/log.txt" into 
file_logging/max_log_files=8
or delete both

File: src/general/OptionsMenu.tscn (hunk 20 above)
Comment (Paulpicotin) [reply src/general/OptionsMenu.tscn:196]: I don't know if this is what you are talking about but I think this problem is when I launch godot to  see the button I make invisible graphics and visible misc and when I leave Godot I forgot to make i back  so now it's done

Comment (Kasterisk) [reply src/general/OptionsMenu.tscn:196]: Yes, the initial tab shown should be Graphics tab.

Comment (Paulpicotin) [reply src/general/OptionsMenu.tscn:196]: I don't know If it's about this you are talking about but I think the problem come's when I leave godot I make invisible graphics and visible misc and now I revert It into a visible graphics and an invisible misc

File: src/general/OptionsMenu.tscn (hunk 21 above)
Comment (Paulpicotin) [reply src/general/OptionsMenu.tscn:798]: That's just appeared when I make invisible misc back

File: project.godot (hunk 10 above)
Comment (hhyyrylainen) [reply project.godot:249]: You need to copy the content that is on master branch so that this PR no longer has any changes shown to project.godot (https://github.com/Revolutionary-Games/Thrive/pull/1785/files)

File: src/general/OptionsMenu.tscn (hunk 20 above)
Comment (hhyyrylainen) [reply src/general/OptionsMenu.tscn:196]: Setting the visibilities back in the godot editor to what they were should resolve this.

File: src/general/OptionsMenu.tscn [hunk 22]
```diff
@@ -938,28 +939,35 @@ text = "USE_A_CUSTOM_USERNAME"
...
 
 [node name="CustomUsername" type="LineEdit" parent="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer/HBoxContainer3"]
-margin_left = 211.0
-margin_right = 511.0
+margin_left = 351.0
+margin_right = 651.0
 margin_bottom = 27.0
 rect_min_size = Vector2( 300, 0 )
 
+[node name="Screeshot Folder" type="Button" parent="CenterContainer/VBoxContainer/PanelContainer/MarginContainer/Misc/VBoxContainer"]
```
Comment (Kasterisk) [inline src/general/OptionsMenu.tscn:964]: The name shouldn't be spaced to make it consistent with other node naming. Also I'd prefer setting the horizontal size flag for the button to none, so it wouldn't become unnecessarily too wide.

**Review by Kasterisk — CHANGES_REQUESTED** [review]

File: src/engine/Settings.cs (hunk 19 above)
Comment (Kasterisk) [reply src/engine/Settings.cs:220]: @Paulpicotin Still unintended changes here, you may want to use `git checkout master src/engine/Settings.cs` to revert these.

**Review by FuJa0815 — APPROVED** [review]
No problems except CI issues and Kasterisk's note

**Comment by hhyyrylainen** [comment]
I'll probably finish this myself later today if no one else does, just so we can get this in the next release.

**Review by hhyyrylainen — APPROVED** [review]
